internal/indexer/documentation.go

```diff
@@ -0,0 +1,748 @@
+	var result funcDocs
+	result.name = in.Name.String()
+	result.exported = ast.IsExported(in.Name.String()) && !isTestFile
+	result.deprecated = isDeprecated(in.Doc.Text())
+	result.docsMarkdown = godocToMarkdown(in.Doc.Text())
+
+	// Create a brand new FuncDecl based on the parts of the input we care about,
+	// ignoring other aspects (e.g. docs and the function body, which are not needed to
+	// produce the function signature.)
+	fn := &ast.FuncDecl{Name: in.Name}
+
+	// Receivers (e.g. struct methods.)
+	if in.Recv != nil {
+		fn.Recv = &ast.FieldList{List: make([]*ast.Field, 0, len(in.Recv.List))}
+		for _, field := range in.Recv.List {
+			fn.Recv.List = append(fn.Recv.List, &ast.Field{
+				Type:  field.Type,
+				Names: field.Names,
+			})
+		}
+		if len(fn.Recv.List) > 0 {
+			// Guaranteed to be length 1 for all valid Go programs, see https://golang.org/ref/spec#Receiver
+			result.recvType = fn.Recv.List[0].Type
+
+			// Mark functions as unexported if they are an exported method of a type that is
+			// unexported.
+			if named, ok := dereference(p.TypesInfo.TypeOf(result.recvType)).(*types.Named); ok {
+				if !named.Obj().Exported() {
+					result.exported = false
+				}
+			}
+		}
+	}
+
+	// Parameters.
+	fn.Type = &ast.FuncType{}
+	fn.Type.Params = &ast.FieldList{List: make([]*ast.Field, 0, len(in.Type.Params.List))}
+	for _, field := range in.Type.Params.List {
+		fn.Type.Params.List = append(fn.Type.Params.List, &ast.Field{
+			Type:  field.Type,
+			Names: field.Names,
+		})
+	}
+
+	// Results.
+	if in.Type.Results != nil {
+		fn.Type.Results = &ast.FieldList{List: make([]*ast.Field, 0, len(in.Type.Results.List))}
+		for _, field := range in.Type.Results.List {
+			fn.Type.Results.List = append(fn.Type.Results.List, &ast.Field{
+				Type:  field.Type,
+				Names: field.Names,
+			})
+			result.resultTypes = append(result.resultTypes, field.Type)
+		}
+	}
+	// TODO(slimsag): future: this doesn't format types very appropriately, some could span
+	// multiple lines!
+	result.label = formatNode(fset, fn)
+	if lines := strings.Split(result.label, "\n"); len(lines) > 1 {
+		result.label = lines[0] + "..." // To be fixed another day!
+	}
+
+	// Produce the full type signature with docs on e.g. anonymous struct fields, multi-line parameters,
+	// etc. but not on the function itself (we'll produce those as Markdown below.)
+	cpy := *in
+	cpy.Doc = nil
+	cpy.Body = nil
+	result.signature = formatNode(fset, &cpy)
+
+	return result
+}
+
+// formatNode turns an ast.Node into a string.
+func formatNode(fset *token.FileSet, node interface{}) string {
+	var buf bytes.Buffer
+	err := format.Node(&buf, fset, node)
+	if err != nil {
+		panic(fmt.Sprintf("never here: %v", err))
+	}
+	return buf.String()
+}
+
+// formatTypeLabel formats the given type as a one-line label, omitting certain details like struct
+// fields.
+func formatTypeLabel(t types.Type) string {
```

Comment:
Can we get specific unit tests for these util functions?

Reply:
Right now this does have quite good test coverage through the `child_symbols.go` + markdown regression/integration tests, but I agree it'd be nice to have unit tests for these util functions as well. I am going to follow up on this and am tracking this here: #20567